File: tests/Web.AcceptanceTests/Pages/LogWorkoutPage.cs
using System.Text.RegularExpressions;

namespace LiftAndShift.Web.AcceptanceTests.Pages;

public class LogWorkoutPage(IPage page) : BasePage(page)
{
    public override string PagePath => $"{BaseUrl}/log-workout";

    public Task AssertHeading() => Assertions.Expect(Page.Locator("h1")).ToHaveTextAsync("Log Workout");

    public async Task SelectExerciseByName(string name)
    {
        var select = Page.Locator("select").First;
        var value = await select.Locator("option", new() { HasText = name }).First.GetAttributeAsync("value");
        await select.SelectOptionAsync(new SelectOptionValue { Value = value });
    }

    public Task ClickAddExercise() => Page.Locator("button:has-text('Add Exercise')").ClickAsync();

    public Task SetFirstSetWeight(string value) =>
        Page.Locator("article table input[type='number']").First.FillAsync(value);

    public Task MarkFirstSetCompleted() =>
        Page.Locator("article table input[type='checkbox']").First.CheckAsync();

    public Task ClickSaveAsDraft() => Page.Locator("button.outline:has-text('Save as Draft')").ClickAsync();

    public Task ClickCompleteWorkout() => Page.Locator("button:has-text('Complete Workout')").ClickAsync();

    public async Task<int> WaitForWorkoutDetailUrlAndGetId()
    {
        await Page.WaitForURLAsync(new Regex(@"/workout/\d+$"));
        var match = Regex.Match(Page.Url, @"/workout/(\d+)$");
        return int.Parse(match.Groups[1].Value);
    }
}

File: src/Web/ClientApp/e2e/pages/log-workout.page.ts
import { expect } from '@playwright/test';
import { BasePage, baseUrl } from './base-page';

export class LogWorkoutPage extends BasePage {
  get pagePath(): string {
    return `${baseUrl()}/log-workout`;
  }

  assertHeading() {
    return expect(this.page.locator('h1')).toHaveText('Log Workout');
  }

  async selectExerciseByName(name: string) {
    const select = this.page.locator('select').first();
    const value = await select.locator('option', { hasText: name }).first().getAttribute('value');
    await select.selectOption(value);
  }

  clickAddExercise() {
    return this.page.locator("button:has-text('Add Exercise')").click();
  }

  setFirstSetWeight(value: string) {
    return this.page.locator("article table input[type='number']").first().fill(value);
  }

  markFirstSetCompleted() {
    return this.page.locator("article table input[type='checkbox']").first().check();
  }

  clickSaveAsDraft() {
    return this.page.locator("button.outline:has-text('Save as Draft')").click();
  }

  clickCompleteWorkout() {
    return this.page.locator("button:has-text('Complete Workout')").click();
  }

  async waitForWorkoutDetailUrlAndGetId(): Promise<number> {
    await this.page.waitForURL(/\/workout\/\d+$/);
    const match = /\/workout\/(\d+)$/.exec(this.page.url());
    return Number(match![1]);
  }
}
